src/storage/garbage_collector.cpp

```diff
@@ -86,80 +120,331 @@ uint32_t GarbageCollector::ProcessUnlinkQueue() {
       // to safely free the txn
       txns_to_deallocate_.push_front(txn);
       txns_processed++;
-    } else if (transaction::TransactionUtil::NewerThan(oldest_txn, txn->TxnId().load())) {
-      // Safe to garbage collect.
+    } else {
+      // This is a txn that may or may not be visible to any running txns. Proceed with unlinking its UndoRecords
+      // with an Interval GC approach
+      bool all_unlinked = true;
       for (auto &undo_record : txn->undo_buffer_) {
-        DataTable *&table = undo_record.Table();
-        // Each version chain needs to be traversed and truncated at most once every GC period. Check
-        // if we have already visited this tuple slot; if not, proceed to prune the version chain.
-        if (visited_slots.insert(undo_record.Slot()).second)
-          TruncateVersionChain(table, undo_record.Slot(), oldest_txn);
-        // Regardless of the version chain we will need to reclaim deleted slots and any dangling pointers to varlens.
-        ReclaimSlotIfDeleted(&undo_record);
-        ReclaimBufferIfVarlen(txn, &undo_record);
+        all_unlinked = all_unlinked && ProcessUndoRecord(&undo_record, &active_txns);
       }
-      txns_to_deallocate_.push_front(txn);
-      txns_processed++;
+      if (all_unlinked) {
+        // We unlinked all of the UndoRecords for this txn, so we can add it to the deallocation queue
+        txns_to_deallocate_.push_front(txn);
+        txns_processed++;
+      } else {
+        // We didn't unlink all of the UndoRecords (UnlinkUndoRecord returned false due to a write-write conflict),
+        // requeue txn for next GC run. Unlinked UndoRecords will be skipped on the next time around since we use the
+        // table pointer of an UndoRecord as the internal marker of being unlinked or not
+        requeue.push_front(txn);
+      }
+    }
+  }
+
+  storage::UndoBuffer *buf = nullptr;
+  std::forward_list<storage::UndoBuffer *> buf_requeue;
```

Comment:
I would suggest that all variables and data structures should be declared at the begining of the function.